src/sql-parser/src/ast/defs/value.rs

```diff
@@ -238,3 +239,211 @@ impl Default for IntervalValue {
+        if let Ok(v) = value.parse::<u64>() {
+            return Ok(Self(v));
+        }
+        let number: String = value
+            .chars()
+            .take_while(|c| c.is_digit(10) || c == &'.')
+            .collect();
+        match number.parse::<f64>() {
+            Ok(v) => {
+                let suffix: String = value
+                    .chars()
+                    .skip_while(|c| c.is_whitespace() || c.is_digit(10) || c == &'.')
+                    .collect();
+                match suffix.parse::<BytesUnit>() {
+                    Ok(u) => Ok(Self(u64::cast_lossy(v * f64::cast_lossy(u.value())))),
+                    Err(error) => Err(format!(
+                        "couldn't parse {:?} into a known SI unit, {}",
+                        suffix, error
+                    )),
+                }
+            }
+            Err(error) => Err(format!(
+                "couldn't parse {:?} into a BytesValue, {}",
+                value, error
+            )),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::ast::BytesValue;
+    use mz_ore::cast::CastLossy;
+
+    #[mz_ore::test]
+    fn test_to_string() {
+        assert_to_string("609.0 PB", BytesValue::pb(609));
+    }
+
+    fn assert_to_string(expected: &str, b: BytesValue) {
+        assert_eq!(expected.to_string(), b.to_string());
+    }
+
+    fn assert_display(expected: &str, b: BytesValue) {
+        assert_eq!(expected, format!("{}", b));
+    }
+
+    #[mz_ore::test]
+    fn test_display() {
+        assert_display("215 B", BytesValue::b(215));
+        assert_display("1.0 KB", BytesValue::kb(1));
+        assert_display("301.0 KB", BytesValue::kb(301));
+        assert_display("419.0 MB", BytesValue::mb(419));
+        assert_display("518.0 GB", BytesValue::gb(518));
+        assert_display("815.0 TB", BytesValue::tb(815));
+        assert_display("609.0 PB", BytesValue::pb(609));
+    }
+
+    #[mz_ore::test]
+    fn test_parse() {
+        // shortcut for writing test cases
+        fn parse(s: &str) -> u64 {
+            s.parse::<BytesValue>().unwrap().0
+        }
+
+        assert_eq!("0".parse::<BytesValue>().unwrap().0, 0);
+        assert_eq!(parse("0"), 0);
+        assert_eq!(parse("500"), 500);
+        assert_eq!(parse("1KB"), BytesValue::kb(1).as_u64());
+        assert_eq!(
+            parse("1.5KB"),
+            u64::cast_lossy(1.5 * f64::cast_lossy(BytesValue::kb(1).as_u64()))
+        );
+        assert_eq!(parse("1 KB"), BytesValue::kb(1).as_u64());
+        assert_eq!(parse("3 MB"), BytesValue::mb(3).as_u64());
+        assert_eq!(parse("6 GB"), BytesValue::gb(6).as_u64());
+        assert_eq!(parse("4GB"), BytesValue::gb(4).as_u64());
+        assert_eq!(parse("88TB"), BytesValue::tb(88).as_u64());
+        assert_eq!(parse("521  TB"), BytesValue::tb(521).as_u64());
+        assert_eq!(parse("8 PB"), BytesValue::pb(8).as_u64());
+        assert_eq!(parse("81 PB"), BytesValue::pb(81).as_u64());
+    }
+}
```

Comment:
Since the SQL parser won't ever parse things as a `BytesValue` directly—it'll always keep them as strings—better to put these types somewhere else! I think `mz_repr` is probably the right place.

Reply:
Moved under `mz_repr`. Also renamed it to `Bytes` (I had done `BytesValue` to follow the pattern for `IntervalValue`).